### PR TITLE
Adds an entity sharing modal

### PR DESCRIPTION
**Pull request by ryan-bradford**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [x] Tests for the changes have been added (for bug fixes / features)
-   [x] Examples have been added / updated (for bug fixes / features)
-   [x] Changelog has been updated

## PR Type

What kind of change does this PR introduce?

-   [ ] Bugfix
-   [x] Feature
-   [ ] Code style update (formatting, local variables)
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Build related changes
-   [ ] CI related changes
-   [ ] Documentation content changes
-   [ ] Example website changes
-   [ ] Version bump
-   [ ] Other... Please describe:

## What does this change do?

Adds the new sharing modal to share entities with individuals. Also adds a directive that is for conveniently configuring the users, groups, orgs use case. 

## Reviewer Notes

1. I chose to implement this without using Form groups or form controls. This is because there are no errors that need to be displayed and the form state is very simple. Should I change this?
2. Technically, tabs use some object that extends Entity, but this felt like a lot of work to not make things much more clear. Should SharingTabs have a generic T that extends Entity?
3. Should entities have an ID or a HREF?
4. Some of the tests for the modal feel like tests for the tab instead of tests for the modal itself. Should I refactor this so the tabs have tests that are moved from the existing ones in the modal?
5. Should we add in a fake item at the end as the warning that we are not displaying all the results? It needs to be a fake item because clarity does not support injecting HTML into the end of the combobox.
6. I chose to use the raw Clarity datagrid because this felt like a complicated use case that our VCD datagrid might not fully cover. Should I refactor to use the vcd datagrid?
7. Should I make the select all placeholder in the tab a reusable component?

## What manual testing did you do?

Ran the modal in the examples app, added users, groups, orgs. Changed users rights. Removed a user. Selected all users and groups. Clicked submit and observed the output. See gif above.

## Screenshots (if applicable)

sharing-modal

## Does this PR introduce a breaking change?

-   [ ] Yes
-   [x] No
-

### CODE REVIEW

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    align-content: center;
+    display: flex;
+}
+
+.content {
+    margin: auto auto auto auto;
+}
+
+.select-all-placeholder {
```
[the remark this answers is not in the document]
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:90]: I was trying to get it to center align within the container. Is there a better way of doing this?

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+import { Observable } from 'rxjs';
+import { ComponentRendererConstructor, ComponentRendererSpec } from '../../datagrid';
+import { NonEmptyArray } from '../sharing-modal.component';
+
+/**
+ * The objects an entity must have. You are allowed to pass a more complicated object.
+ */
+export interface Entity {
+    /**
+     * The name of this entity.
```
[the remark this answers is not in the document]
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.ts:18]: I actually just removed the name property all together. There isn't a reason that this interface needs to have a Name.

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    /**
+     * The entities currently selected in the combobox,
+     */
+    comboboxSelection: Entity[] = [];
+
+    /**
+     * Called when the user types something into or opens the combobox. Returns the results from that search.
+     */
+    @Input()
+    makeSearch = (criteria?: string): Promise<SearchResult> => {
```
[the remark this answers is not in the document]
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.ts:179]: Changed to throw an error. Unsure about the second part...

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:90]: The select-all-placeholder is an icon that is horizontally aligned within a div that is center aligned within a rectangle that appears as a gray box.

**Comment by ryan-bradford** [comment]
Fixed all the comments. Also, figured a way to display the search items warning using the content projection that we now use to display the error. Screenshots:

Screen Shot 2021-02-01 at 2 30 04 PM
Screen Shot 2021-02-01 at 2 30 12 PM

File: package.json [hunk 4]
```diff
@@ -35,10 +35,10 @@
     "@angular/platform-browser": "10.2.3",
     "@angular/platform-browser-dynamic": "10.2.3",
     "@angular/router": "10.2.3",
-    "@clr/angular": "4.0.9",
-    "@clr/core": "4.0.9",
-    "@clr/icons": "4.0.9",
-    "@clr/ui": "4.0.9",
+    "@clr/angular": "4.0.10",
```
Comment (juanmendes) [inline package.json:38]: Do we plan on upgrading vcd_ui? Do we need to?

File: projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    rightsOptions: ComboOption[];
+
+    /**
+     * The text that is displayed next to the toggle.
+     */
+    @Input()
+    description: LazyString;
+
+    /**
+     * Ouput when the value behind select all is changed or undefined if select all is disabled.
```
Comment (juanmendes) [inline projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts:36]: What is the value that is passed to the listener?

File: projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * A togggle that allows you to select all and choose the type of right to apply.
+ */
+@Component({
+    selector: 'vcd-select-all-toggle',
+    templateUrl: 'select-all-toggle.component.html',
+    styleUrls: ['./select-all-toggle.component.scss'],
+})
+export class VcdSelectAllToggleComponent {
```
Comment (juanmendes) [inline projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts:22]: Shouldn't this be a control value accessor like the rest of our form controls?

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss [hunk 7]
```diff
@@ -0,0 +1,116 @@
+::ng-deep .clr-form-control {
+    margin-top: 0 !important;
```
Comment (juanmendes) [inline projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:2]: Pleas avoid `!important`. Add specificity, if possible

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+clr-datagrid {
```
Comment (juanmendes) [inline projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:5]: Why is this needed instead of just a height?

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    display: flex;
+    margin-top: 20px;
+}
+
+button.add {
+    margin-right: 0;
+    flex: 0;
+}
+
+.search-container ::ng-deep .clr-select-wrapper {
```
Comment (juanmendes) [inline projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:20]: Why are you going deep into another component?

File: package.json (hunk 4 above)
Comment (ryan-bradford) [reply package.json:38]: I was trying to see if the Clarity bug with the multiselect combobox was fixed (it wasn't). I'm going to keep the version bumped, but no reason to bump it in VCD_UI. Although, I am going to update package.json to say we require Clarity v4

File: projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss (hunk 8 above)
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/tabs/sharing-modal-tab.component.scss:5]: These styles were no longer needed anymore, so I removed them.

File: projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts (hunk 6 above)
Comment (ryan-bradford) [reply projects/components/src/sharing-modal/select-all-checkbox/select-all-toggle.component.ts:22]: We talked and decided that this is fine left as not a form control because it is not a reusable component.

**Review by juanmendes — APPROVED** [review]
